### PR TITLE
Add commit signing repo rules warning

### LINKED ISSUE #17309: Unable to commit with Signed Commit Rule and SSH
### The problem

Hello, I just upgraded GH Desktop, but now can no longer commit.
We have Rulesets that require signed commits, and have SSH setup for verified commits.
Before the upgrade today, all was working. Now I get an error message in Desktop preventing commits.

Please see screenshots / gitconfig below.

### Release version

3.3.0 (arm64)

### Operating system

macOS Ventura 13.5

### Steps to reproduce the behavior

Setup SSH on macOS and Github.
Add GH Ruleset to require Commit Signature.
Try to commit using Desktop.
Unable to commit even though SSH signed commits is active, and verified by previous commits.

### Log files

_No response_

#### Error:
desktop-error

#### Rulesets:
rulesets-signed-commits

#### SSH Commits:
ssh-commits

#### .gitconfig
```
[user]
  name = <name>
  email = <email>
  signingKey = ~/.ssh/id_ed25519.pub

[credential]
  helper = osxkeychain

[gpg]
  format = ssh

[commit]
  gpgSign = true
```

### PR DESCRIPTION
**Pull request by vaindil**
Closes #17309

## Description
Commit signing was not previously being checked specially for repo rules, it was bundled into the basic commit warning. This breaks it out so it's properly checked.

This PR adds a new warning/error that's only displayed if the user's `.gitconfig` does not have some truthy variant of `commit.gpgsign`, [per our docs](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits).

### Screenshots

New commit signing error message

New commit signing warning message when rules can be bypassed

## Release notes

Notes: Added a check for the "require signed commits" repo rule

### CODE REVIEW

File: app/src/lib/helpers/repo-rules.ts
Comment (tidy-dev) [inline app/src/lib/helpers/repo-rules.ts]:
```suggestion
  let gitConfigKeySet: boolean | null = null
```

It doesn't seem like this is needed to be defined before hand, and should be able to in the switch statement replace the first `if` with `const gitConfigKeySet = ...`

**Review by tidy-dev — DISMISSED** [review]
✨  I tested and it all works. I have one comment to change in the code.

File: app/src/lib/helpers/repo-rules.ts
Comment (vaindil) [reply app/src/lib/helpers/repo-rules.ts]: I declared it up there to cache it across rule checks--if there are multiple rules that require signed commits, there's no need to query the git config more than once because it's always the same repo. Tiny optimization but I figured it made sense.

Comment (tidy-dev) [reply app/src/lib/helpers/repo-rules.ts]: That makes sense. However, I think since currently it is only one, we would prefer to keep in with our engineering value about maintaining [Immutability](https://github.com/desktop/desktop/blob/development/docs/contributing/engineering-values.md#immutability-is-good). 

In the future if it is needed, we can likely do something like memoize for this sort of optimization. :)

Comment (vaindil) [reply app/src/lib/helpers/repo-rules.ts]: Sorry, I didn't explain properly. Any number of rulesets can apply to the user's branch, which can result in having multiple of this type of rule. That's why I put this in here--this case may be hit multiple times right now.

Whether the underlying git operation is slow enough to justify this optimization against the code style is another question, but it is at least a current optimization, it's not future-proofing.

File: app/src/lib/helpers/repo-rules.ts
Comment (niik) [inline app/src/lib/helpers/repo-rules.ts]: If the key isn't found here `getBooleanConfigValue` will return null which will cause this to exec the next loop as well, negating the purpose of the `gitConfigKeySet` perf optimization cache.

If `getBoolConfigValue` returns null we should store the Git default

**Review by niik — CHANGES_REQUESTED** [review]
Me and @sergiou87 took a look at this and given that :flag-us: is asleep now I'll make the suggested changes and have @sergiou87 re-review before we merge and push a hotfix.

One thing that stood out to the both of us is that the check for the `commit.gpgsign` config value happens during rule parsing time and not during enforcement. We wondered what would happen if a user gets blocked by the signed commits rule, goes to the terminal to flip the `commit.gpgsign` config and then goes back to the app. Would the rule be dismissed then? Do we parse repo rules when the app regains focus?

**Review by sergiou87 — APPROVED** [review]

**Review by niik — APPROVED** [review]

File: app/src/lib/helpers/repo-rules.ts
Comment (vaindil) [reply app/src/lib/helpers/repo-rules.ts]: 🤦 Good point, thank you for fixing it!

**Pull request merged**
